runtime/doc/lua.txt

```diff
@@ -3205,22 +3211,26 @@ Lua module: vim.iter                                                *vim.iter*
 *vim.iter()* is an interface for |iterable|s: it wraps a table or function
 argument into an *Iter* object with methods (such as |Iter:filter()| and
 |Iter:map()|) that transform the underlying source data. These methods can
-be chained together to create iterator "pipelines". Each pipeline stage
-receives as input the output values from the prior stage. The values used
-in the first stage of the pipeline depend on the type passed to this
-function:
-
-• List tables (arrays) pass only the value of each element
-• Non-list tables (dictionaries) pass both the key and value of each
-  element
-• Function |iterator|s pass all of the values returned by their respective
-  function
-• Tables with a metatable implementing |__call()| are treated as function
-  iterators
-
-The iterator pipeline terminates when the original table or function
-iterator runs out of values (for function iterators, this means that the
-first value returned by the function is nil).
+be chained to create iterator "pipelines": the output of each pipeline
+stage is input to the next stage. The first stage depends on the type
+passed to `vim.iter()`:
+
+• List tables (arrays, |lua-list|) yield only the value of each element.
+  • Use |Iter:enumerate()| to also pass the index to the next stage.
+  • Or initialize with ipairs(): `vim.iter(ipairs(…))`.
+
+• Non-list tables (|lua-dict|) yield both the key and value of each
+  element.
+• Function |iterator|s yield all values returned by the underlying
+  function.
+• Tables with a |__call()| metamethod are treated as function iterators.
+
+The iterator pipeline terminates when the underlying |iterable| is
+exhausted (for function iterators this means it returned nil).
+
+Note: `vim.iter()` scans table input to decide if it is a list or a dict
+(see |list-iterator|). To avoid this cost, wrap the table in an iterator
+such as ipairs(): `vim.iter(ipairs({…}))`.
```

Comment:
`ipairs()` isn't a perfect solution because it precludes using any iterator methods that operate from the back of the list (e.g. `:rev()`, `:skipback()`, etc.)

In #24746 I proposed exposing `ListIter` directly or providing some kind of "I know what I'm doing" escape hatch for cases where users want to avoid the scan. I still think we could do something like this.

Reply:
Noted in latest push.